### PR TITLE
Transition to executing when begin planning

### PR DESCRIPTION
**Pull request by JunhyungSong**
## Description
Currently, `/v1/statement/queued/{queryId}/{slug}/{token}` get request returns queued query state even if the query is already planning. It will not be redirected to `/v1/statement/executing/{queryId}/{slug}/{token}` get request until all planning tasks including plan distribution is completed. When planning takes long time or fails during planning, it can mislead callers because of this. So, it needs to set `submitted` when transitioning to `PLANNING`.

## Release notes

(v) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by findepi** [comment]
@JunhyungSong thanks for a very nice PR description, this is very useful!

> Currently, `/v1/statement/queued/{queryId}/{slug}/{token}` get request returns queued query state even if the query is already planning. It will not be redirected to `/v1/statement/executing/{queryId}/{slug}/{token}` get request until all planning tasks including plan distribution is completed. When planning takes long time or fails during planning, it can mislead callers because of this.

Callers should not infer any particular meaning from the paths used in nextUri. 

Callers (client tools, UIs), should inspect the query state, so we should make sure the query state returned as "PLANNING" if the query is in the planning state.

> So, it needs to set submitted when transitioning to PLANNING.

What is the effect of this (from the query protocol perspective)?

**Comment by JunhyungSong** [comment]
> What is the effect of this (from the query protocol perspective)?

This will mark `dispatched` as `true`. And it will set coordinator location and change the next URI to `ExecutingStatementResource` instead of `QueuedStatementResource`. With that, `ExecutingStatementResource` will return `QueryResult` with a correct state`(e.g. `PLANNING` or `STARTING` instead of `QUEUED`.

**Comment by findepi** [comment]
>  will return `QueryResult` with a correct state(e.g. `PLANNING` or `STARTING` instead of `QUEUED`).

that's sounds like something observable from end user perspective.

So, are you saying that queries in PLANNING state are not viewed as such by the client tools?
would it be possible to write a test for that?
for example with a mock connector you can insert a pause in some metadata method, allowing planning to take as much time as needed for test purposes.

**Comment by JunhyungSong** [comment]
> So, are you saying that queries in PLANNING state are not viewed as such by the client tools?

Correct.

> would it be possible to write a test for that?
for example with a mock connector you can insert a pause in some metadata method, allowing planning to take as much time as needed for test purposes.

Can you recommend tests that I can refer to?

**Comment by findepi** [comment]
I do not. @electrum and @dain will know better for protocol-level changes and tests

**Comment by JunhyungSong** [comment]
@electrum @dain do you have any ideas on how to add tests for this?

**Comment by bitsondatadev** [comment]
@pettyjamesm would you also be able to review this one or approve if you've already done a review internally please?

File: core/trino-main/src/main/java/io/trino/server/protocol/Query.java [hunk 1]
```diff
@@ -459,18 +459,24 @@ private synchronized QueryResults getNextResult(long token, UriInfo uriInfo, Dat
         QueryInfo queryInfo = queryManager.getFullQueryInfo(queryId);
         queryManager.recordHeartbeat(queryId);
 
-        closeExchangeIfNecessary(queryInfo);
```
Comment (pettyjamesm) [inline core/trino-main/src/main/java/io/trino/server/protocol/Query.java:462]: I think `closeExchangeIfNecessary(QueryInfo)` needs to check `queryInfo.getState().ordinal() > STARTING` logic too, otherwise we could close the exchange before planning completes if that were to be called from another code path.

After that change, we should leave the `closeExchangeIfNecessary` and `removePagesFromExchange` as-is without needing to check `isStarted` at this point explicitly.

File: core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java [hunk 2]
```diff
@@ -137,25 +137,24 @@ private void waitForMinimumWorkers()
...
-                }
-                finally {
-                    submitted.set(null);
+        if (stateMachine.transitionToDispatching()) {
+            try {
+                querySubmitter.accept(queryExecution);
+                if (notificationSentOrGuaranteed.compareAndSet(false, true)) {
+                    queryExecution.addFinalQueryInfoListener(queryMonitor::queryCompletedEvent);
                 }
+                queryExecutor.execute(() -> queryExecution.start());
```
[the remark this answers is not in the document]
Comment (JunhyungSong) [reply core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java:146]: Talked offline. DispatchExecutor will not throw RejectedExecutionException unless the server is shutting down.

File: core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java [hunk 3]
```diff
@@ -123,8 +123,8 @@ private void waitForMinimumWorkers()
             }
             ListenableFuture<Void> minimumWorkerFuture = clusterSizeMonitor.waitForMinimumWorkers(executionMinCount, getRequiredWorkersMaxWait(session));
             // when worker requirement is met, start the execution
-            addSuccessCallback(minimumWorkerFuture, () -> startExecution(queryExecution));
-            addExceptionCallback(minimumWorkerFuture, throwable -> queryExecutor.execute(() -> stateMachine.transitionToFailed(throwable)));
+            addSuccessCallback(minimumWorkerFuture, () -> startExecution(queryExecution), queryExecutor);
+            addExceptionCallback(minimumWorkerFuture, throwable -> stateMachine.transitionToFailed(throwable), queryExecutor);
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java:127]: These seem like unrelated changes.  This moves the call back from a direct executor to an explicit executor, which may be useful for some case, but that seems to be unrelated to what is the core of this commit.

File: core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java [hunk 4]
```diff
@@ -365,8 +365,8 @@ public BasicQueryInfo getBasicQueryInfo()
     @Override
     public void start()
     {
-        try (SetThreadName ignored = new SetThreadName("Query-%s", stateMachine.getQueryId())) {
-            try {
+        try {
+            try (SetThreadName ignored = new SetThreadName("Query-%s", stateMachine.getQueryId())) {
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:369]: Why change this?  Also this seems unrelated, so I'd move it to a different commit if it is needed

File: core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java (hunk 2 above)
Comment (dain) [inline core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java:146]: I think this breaks the abstraction between `LocalDispatchQuery` and `SqlQueryManager`.  The SqlQueryManager has been responsible for coordinating the start of queries after and making sure that the interactions with the query tracker are handled property. 

I also don't understand why moving the start here helps with the root issue of reporting query state.

Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java:146]: The issue is that by calling `queryExecution.start()` synchronously, the `submitted` future won't be completed until _after_ it returns which means that it won't happen until after planning completes. That future controls when the client will be redirected to the `ExecutingStatementResource` from `QueuedStatementResource` which prevents the client from observing `PLANNING` as a state. The other reordering is incidental but necessary to ensure that the callbacks are all registered appropriately before the potential failure of the query execution could occur (likely from inside `start()`).

File: core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java (hunk 3 above)
Comment (JunhyungSong) [reply core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java:127]: Right. This is not directly related to this PR. I thought it would be small enough to be included in this PR as a small clean-up. However, if you think that it would be better to have a separate PR, I will exclude this change from this PR. Please let me know.

File: core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java (hunk 2 above)
Comment (JunhyungSong) [reply core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java:146]: @dain I think it doesn't necessarily break the abstraction. Because we can say that query submission is for preparing query execution(e.g. createQuery) and `queryExecution.start()` is for starting query execution. However, if you prefer the current level of abstraction, I can change SqlQueryManager instead to call `start()` using executor. But, with that, as James mentioned, `queryExecution.addFinalQueryInfoListener` will be added after the asynchronous `start()` call which would be more appropriate to be added before `start()`. Let me know what you think.

File: core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java (hunk 4 above)
Comment (JunhyungSong) [reply core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:369]: Previously, if it throws an exception(regardless of likelihood), previously it was caught by LocalDispatchQuery since `start()` was synchronous. However, after this change which makes `start()` asynchronous, the exception will not be caught properly, so the query will not be failed properly. If you want, I will move this to a separate commit. Or do you want to have a separate PR?

**Comment by JunhyungSong** [comment]
The test failures seem irrelevant.

File: core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java (hunk 4 above)
Comment (dain) [reply core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java:375]: Add a comment explaining this

File: core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java [hunk 5]
```diff
@@ -137,25 +138,30 @@ private void waitForMinimumWorkers()
...
-                finally {
+            }
+            catch (Throwable t) {
+                // this should never happen but be safe
+                stateMachine.transitionToFailed(t);
+                log.error(t, "query submitter threw exception");
+                throw t;
+            }
+            finally {
+                if (!submitted.isDone()) {
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java:160]: Why add the `if` statement? I believe that `set` on a done future is a noop.  Also the future could complete between the `isDone` check and `set`

File: core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java [hunk 6]
```diff
@@ -137,25 +138,30 @@ private void waitForMinimumWorkers()
...
     private void startExecution(QueryExecution queryExecution)
     {
-        queryExecutor.execute(() -> {
-            if (stateMachine.transitionToDispatching()) {
-                try {
-                    querySubmitter.accept(queryExecution);
-                    if (notificationSentOrGuaranteed.compareAndSet(false, true)) {
-                        queryExecution.addFinalQueryInfoListener(queryMonitor::queryCompletedEvent);
-                    }
+        if (stateMachine.transitionToDispatching()) {
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java:141]: Is moving the `if` statement outside of the `try` block significant?

File: core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java
Comment (dain) [inline core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java:1]: I think I finally understand what you are trying to do here (correct me if I'm wrong)....

You want to mark the query as submitted as soon as the query begins planning.  This is important because the protocol uses the submitted flag to control access to resources such as query state.  The existing code marks the query as submitted after sending it to the `Consumer<QueryExecution> querySubmitter`, and since that code is single threaded, the submitted flag it not set until after the entire query is planned.

I think there is a much easier way to accomplish this.  The query state machine has event listeners for when the query changes, and we can hook into that so when the query moves to planning (or some deeper state), we mark it as submitted.  We already have such a listener setup in the constructor of this class, but it only marks the query as submitted for done states.

I think you can implement this change by, changing this code in constructor:
```java
            if (state.isDone()) {
                submitted.set(null);
                queryExecutionFuture.cancel(true);
            }
```
To :
```java
            // any PLANNING or later state means the query has been submitted for execution
            if (state.ordinal() >= QueryState.PLANNING.ordinal()) {
                submitted.set(null);
            }
            if (state.isDone()) {
                queryExecutionFuture.cancel(true);
            }
```

**Review by dain — COMMENTED** [review]
For the `Transition to executing when begin planning` commit, it is really important for me to be able to understand the diff if the commit contains only the changes absolutely necessary for the desired change to the state transitions (the part of the code base is really complex).  The other minor changes can be in commits before of after that one.

**Comment by JunhyungSong** [comment]
The test failure seems irrelevant.

File: testing/trino-tests/src/test/java/io/trino/tests/TestLocalDispatchQuery.java [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+                Slug.createNew(),
+                TestingSessionContext.fromSession(TEST_SESSION),
+                "SELECT * FROM lineitem",
+                Optional.of(new TestingQueryExecution()),
+                false)
+                .get();
+
+        // wait until query starts running
+        countDownLatch.await();
+        assertTrue(localDispatchQuery.getDispatchInfo().getCoordinatorLocation().isPresent());
```
Comment (dain) [inline testing/trino-tests/src/test/java/io/trino/tests/TestLocalDispatchQuery.java:89]: Does this test pass in this commit?  If not, please move merge it with the next commit, where the change is implemented

File: core/trino-main/src/main/java/io/trino/server/protocol/Query.java [hunk 8]
```diff
@@ -463,18 +463,24 @@ private synchronized QueryResults getNextResult(long token, UriInfo uriInfo, Dat
...
-
-        // fetch result data from exchange
-        QueryResultRows resultRows = removePagesFromExchange(queryInfo, targetResultSize.toBytes());
+        boolean isStarted = queryInfo.getState().ordinal() > QueryState.STARTING.ordinal();
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/server/protocol/Query.java:466]: Is this change unrelated to the `submitted` change above?  If so, can you move this to another commit and explain why the change is needed (I don't understand why this change is being made)

**Review by dain — COMMENTED** [review]
I'm still having some trouble understanding some of the additional changes in this.  Also, I think the main commit can be simplified by rewriting the test.

File: testing/trino-tests/src/test/java/io/trino/tests/TestLocalDispatchQuery.java [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+    }
+
+    private static class TestingQueryExecution
```
[the remark this answers is not in the document]
Comment (JunhyungSong) [reply testing/trino-tests/src/test/java/io/trino/tests/TestLocalDispatchQuery.java:92]: Ok. Will try.

File: testing/trino-tests/src/test/java/io/trino/tests/TestLocalDispatchQuery.java (hunk 7 above)
Comment (JunhyungSong) [reply testing/trino-tests/src/test/java/io/trino/tests/TestLocalDispatchQuery.java:89]: I wanted to check if the test fails without the fix. Anyways, I have no objections to change the order of the commits. Will move it.

File: core/trino-main/src/main/java/io/trino/server/protocol/Query.java (hunk 8 above)
Comment (JunhyungSong) [reply core/trino-main/src/main/java/io/trino/server/protocol/Query.java:466]: This is related. Without this, when `getQueryResult` is called during planning, it will call `closeExchangeIfNecessary` and `removePagesFromExchange`, and then the query will get stuck. Previously, this function was not called because `getQueryResult` returned a dummy result through `QueuedStatementResource`.

File: core/trino-main/src/main/java/io/trino/dispatcher/DispatchManager.java [hunk 10]
```diff
@@ -76,6 +75,8 @@
 
     private final QueryManagerStats stats = new QueryManagerStats();
 
+    private DispatchQueryFactory dispatchQueryFactory;
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/dispatcher/DispatchManager.java:78]: We should not be making this field mutable for a test.  I'll explain how in the test code below.

**Comment by JunhyungSong** [comment]
The failure looks irrelevant.
